controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -88,17 +88,72 @@ func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, request rec
 
 	// if no expansion is in-progress update the phase to "not-started"
 	if storageAutoScaler.Status.Phase == "" {
-		originalStorageAutoScaler := storageAutoScaler.DeepCopy()
 		storageAutoScaler.Status.Phase = ocsv1.StorageAutoScalerPhaseNotStarted
-		err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+		err := r.updateStatus(ctx, storageAutoScaler)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 	}
 
-	// if the expansion is InProgress or Failed, verify the scaling
-	if storageAutoScaler.Status.Phase == ocsv1.StorageAutoScalerPhaseInProgress || storageAutoScaler.Status.Phase == ocsv1.StorageAutoScalerPhaseFailed {
-		return r.verifyScaling(ctx, storageAutoScaler)
+	// get the current osd size and count from prometheus
+	currentOsdSize, currentOsdCount, err := r.getCurrentOsdSizeAndCount(ctx, storageAutoScaler)
+	if err != nil {
+		r.Log.Error(err, "failed to get current osd size and count")
+		return reconcile.Result{}, err
+	}
+
+	// get the desired osd size and count from storagecluster
+	desiredOsdSize, desiredOsdCount := getDesiredOsdSizeAndCount(storageCluster, storageAutoScaler)
+
+	// expansion is InProgress
+	if currentOsdSize.Cmp(desiredOsdSize) != 0 || currentOsdCount != desiredOsdCount {
+		if timeoutHasElapsed(storageAutoScaler.Spec.TimeoutSeconds, storageAutoScaler.Status.LastExpansion) {
+			storageAutoScaler.Status.Phase = ocsv1.StorageAutoScalerPhaseFailed
+			storageAutoScaler.Status.Error = &ocsv1.TimestampedError{
+				Message:   "scaling verification timed out for storageAutoScaler " + storageAutoScaler.Name + "device class " + storageAutoScaler.Spec.DeviceClass,
+				Timestamp: metav1.Now(),
+			}
+			err := r.updateStatus(ctx, storageAutoScaler)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			err = fmt.Errorf("scaling verification timed out, for device class %s", storageAutoScaler.Spec.DeviceClass)
+			r.Log.Error(err, "scaling verification timed out")
```

Comment:
Do we need to return here after logging the err?

Reply:
We don't need to return error, 
As we will still try to scale even if the scaling is failed after `timeout`.

Reply:
Sounds good to continue attempting. I'd suggest a different error message. "timed out" implies that the request is canceled, but we are still continuing so I'd suggest not calling it "timed out".
```suggestion
			r.Log.Error(err, "auto scaling is taking longer than expected and may not succeed")
```

Reply:
Done